File: 03/03-merge-two-sorted-lists.py
# https://leetcode.com/problems/merge-two-sorted-lists/

from typing import Optional

class ListNode(object):
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

    def __str__(self) -> str:
        return f'{self.val}, {self.next}'

def printSolution(node: ListNode):
    print(f'[{node}]')

class Solution:

    def isNodeValid(self, node: ListNode):
        return True if node and node.val is not None else False

    def mergeTwoLists(self, list1: Optional[ListNode], list2: Optional[ListNode]) -> Optional[ListNode]:
        final_list = ListNode()
        head = final_list
        while (self.isNodeValid(list1) or self.isNodeValid(list2)):
            print(list2 and list2.val)
            if (self.isNodeValid(list1) and self.isNodeValid(list2) and list1.val < list2.val
               or self.isNodeValid(list1) and not self.isNodeValid(list2)):
                final_list.next = ListNode(list1.val)
                list1 = list1.next
            elif (self.isNodeValid(list2)):
                final_list.next = ListNode(list2.val)
                list2 = list2.next
            final_list = final_list.next

        return head.next


list1 = ListNode(1, ListNode(4, ListNode(7)))
list2 = ListNode(5, ListNode(8))
# printSolution(Solution().mergeTwoLists(list1, list2))
# printSolution(Solution().mergeTwoLists(ListNode(), ListNode()))
# printSolution(Solution().mergeTwoLists(ListNode(None), ListNode(0)))
# printSolution(Solution().mergeTwoLists(ListNode(0), ListNode(None)))

File: 03-merge-two-sorted-lists/03-merge-two-sorted-lists.js
// https://leetcode.com/problems/merge-two-sorted-lists/

import { logObject } from '../shared/utils.js';
import { ListNode } from './common/list-node.model.js';

/**
 * @param {ListNode} node
 * @returns node valid
 */
function isNodeValid(node) {
    if (node && node.val !== null) {
        return true;
    }
    return false;
}

/**
 * @param {ListNode} list1
 * @param {ListNode} list2
 * @return {ListNode}
 */
var mergeTwoLists = function (list1, list2) {
    let finalList = new ListNode();
    const head = finalList;

    while (isNodeValid(list1) || isNodeValid(list2)) {
        if (
            (isNodeValid(list1) && isNodeValid(list2) && list1.val < list2.val) ||
            (isNodeValid(list1) && !isNodeValid(list2))
        ) {
            finalList.next = new ListNode(list1.val);
            list1 = list1.next;
        } else if (isNodeValid(list2)) {
            finalList.next = new ListNode(list2.val);
            list2 = list2.next;
        }
        finalList = finalList.next;
    }
    return head.next;
};

const list1 = new ListNode(1, new ListNode(4, new ListNode(7)));
const list2 = new ListNode(5, new ListNode(8));
logObject(mergeTwoLists(list1, list2));
// logObject(mergeTwoLists(new ListNode(), new ListNode()));
// logObject(mergeTwoLists(new ListNode(null), new ListNode(0)));
// logObject(mergeTwoLists(new ListNode(0), new ListNode(null)));
